File: samples/ble_util/src/com/idevicesinc/sweetblue/ble_util/DeviceListEntry.java
package com.idevicesinc.sweetblue.ble_util;

import android.content.Context;
import android.text.SpannableString;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.idevicesinc.sweetblue.*;
import com.idevicesinc.sweetblue.BleDevice.ConnectionFailListener.Info;
import com.idevicesinc.sweetblue.BleDevice.ConnectionFailListener.Please;
import com.idevicesinc.sweetblue.utils.Interval;
import com.idevicesinc.sweetblue.utils.Utils;
import com.idevicesinc.sweetblue.utils.State.ChangeIntent;

/**
 * 
 * @author dougkoellmer
 */
public class DeviceListEntry extends FrameLayout implements BleDevice.StateListener
{	
	private final BleDevice m_device;
	private final Button m_connect;
	private final Button m_disconnect;
	private final Button m_bond;
	private final Button m_unbond;
	private final TextView m_status;
	private final TextView m_name;
	
	public DeviceListEntry(Context context, BleDevice device)
	{
		super(context);
		
		m_device = device;
		m_device.setListener_State(this);
		m_device.setListener_ConnectionFail(new BleDevice.DefaultConnectionFailListener()
		{
			@Override public Please onConnectionFail(Info moreInfo)
			{
				Please please = super.onConnectionFail(moreInfo);
				
				if( please == Please.DO_NOT_RETRY )
				{
					Toast.makeText(getContext(), moreInfo.device().getName_debug() + " connection failed with " + moreInfo.failureCountSoFar() + " retries - " + moreInfo.reason(), Toast.LENGTH_LONG).show();
				}
				
				return please;
			}
		});
		
		LayoutInflater li = LayoutInflater.from(context);
		View inner = li.inflate(R.layout.device_entry, null);
		
		m_connect = (Button) inner.findViewById(R.id.connect_button);
		m_disconnect = (Button) inner.findViewById(R.id.disconnect_button);
		m_bond = (Button) inner.findViewById(R.id.bond_button);
		m_unbond = (Button) inner.findViewById(R.id.unbond_button);
		m_status = (TextView) inner.findViewById(R.id.device_status);
		m_name = (TextView) inner.findViewById(R.id.device_name);
		
		m_connect.setOnClickListener(new OnClickListener()
		{
			@Override public void onClick(View v)
			{
				m_device.connect();
			}
		});
		
		m_disconnect.setOnClickListener(new OnClickListener()
		{
			@Override public void onClick(View v)
			{
				m_device.disconnect();
			}
		});
		
		m_bond.setOnClickListener(new OnClickListener()
		{
			@Override public void onClick(View v)
			{
				m_device.bond();
			}
		});
		
		m_unbond.setOnClickListener(new OnClickListener()
		{
			@Override public void onClick(View v)
			{
				m_device.unbond();
			}
		});

		updateStatus(m_device.getStateMask());
		
		String name = m_device.getName_normalized();
		if( name.length() == 0 )
		{
			name = m_device.getMacAddress();
		}
		else
		{
			name += "(" + m_device.getMacAddress() + ")";
		}
		m_name.setText(name);
		
		this.addView(inner);
		
		if( device.getLastDisconnectIntent() == ChangeIntent.UNINTENTIONAL )
		{
			device.connect();
		}
	}
	
	public BleDevice getDevice()
	{
		return m_device;
	}
	
	private void updateStatus(int deviceStateMask)
	{
		SpannableString status = Utils.makeStateString(BleDeviceState.values(), deviceStateMask);
		m_status.setText(status);
	}

	@Override public void onStateChange(ChangeEvent event)
	{
		updateStatus(event.newStateBits());
	}
}

File: xamarin/SweetBlue/component/samples/Android/BleUtil/BleUtil/DeviceListEntry.cs
using System;
using Android.Views;
using Android.Widget;
using Android.Content;
using Android.Text;
using Android.Util;
using Idevices.Sweetblue;
using Idevices.Sweetblue.Util;


namespace BleUtil
{
	public class DeviceListEntry : FrameLayout
	{

		private BleDevice m_device;
		private Button m_connect;
		private Button m_disconnect;
		private Button m_bond;
		private Button m_unbond;
		private TextView m_status;
		private TextView m_name;

		public DeviceListEntry (Context context, BleDevice device) : base(context)
		{
			m_device = device;
			m_device.State += (sender, e) => {				
				updateStatus(e.StateEvent.NewStateBits());
			};

			m_device.SetListener_ConnectionFail(new Fail (context));

			m_device.Bonded += (sender, args) => {				
				String toast = args.BondEvent.Device().Name_debug + " bond attempt finished with status " + args.BondEvent.Status();
				Toast.MakeText(context, toast, ToastLength.Long).Show();
			};

			View inner = LayoutInflater.From (context).Inflate (Resource.Layout.Device_Entry, null);
			m_connect = inner.FindViewById<Button> (Resource.Id.connect_button);
			m_disconnect = inner.FindViewById<Button> (Resource.Id.disconnect_button);
			m_bond = inner.FindViewById<Button> (Resource.Id.bond_button);
			m_unbond = inner.FindViewById<Button> (Resource.Id.unbond_button);
			m_status = inner.FindViewById<TextView> (Resource.Id.device_status);
			m_name = inner.FindViewById<TextView> (Resource.Id.device_name);

			m_connect.Click += delegate {
				m_device.Connect();
			};

			m_disconnect.Click += delegate {
				m_device.Disconnect();
			};

			m_bond.Click += delegate {
				m_device.Bond();
			};

			m_unbond.Click += delegate {
				m_device.Unbond();
			};

			updateStatus (m_device.StateMask);

			String name = m_device.Name_normalized;
			if (name.Length == 0) {
				name = m_device.MacAddress;
			} else {
				name += "(" + m_device.MacAddress + ")";
			}
			m_name.Text = name;

			AddView (inner);

			if (device.LastDisconnectIntent == StateChangeIntent.Unintentional) {
				device.Connect ();
			}
		}

		public BleDevice getDevice() {
			return m_device;
		}

		private void updateStatus(int deviceStateMask) {
			SpannableString status = Utils_String.MakeStateString (BleDeviceState.Values(), deviceStateMask);
			m_status.TextFormatted = status;
		}


		public class Fail : BleDevice.DefaultConnectionFailListener {

			private Context m_context;

			public Fail(Context context) : base()
			{
				m_context = context;
			}

			public override BleDevice.ConnectionFailListenerPlease OnEvent (BleDevice.ConnectionFailListenerConnectionFailEvent ev)
			{
				BleDevice.ConnectionFailListenerPlease please = base.OnEvent (ev);
				if (!please.IsRetry) {					
					String toast = ev.Device ().Name_debug + " connection failed with " + ev.FailureCountSoFar() + " retries - " + ev.Status ().ToString();
					Toast.MakeText (m_context, toast, ToastLength.Long).Show ();

				}
				return please;
			}
		}

	}
}

